core/internal/runmetric/runmetric.go

```diff
@@ -115,21 +115,23 @@ func (mh *MetricHandler) UpdateMetrics(
 	return mh.createGlobMetrics(history)
 }
 
-// InsertStepMetrics inserts an automatic step metric for every defined
-// metric with step_sync set to true.
+// InsertStepMetrics inserts an automatic step metric for every metric
+// that has a step metric defined and is not already present in the
+// history.
 func (mh *MetricHandler) InsertStepMetrics(
 	history *runhistory.RunHistory,
 ) {
-	for key, metricDef := range mh.definedMetrics {
-		if len(key) == 0 {
-			continue
+	history.ForEachKey(func(path pathtree.TreePath) bool {
+		key := strings.Join(path.Labels(), ".")
+		metricDef, ok := mh.definedMetrics[key]
+		// This should never happen, but we'll skip the metric if it does.
```

Comment:
this comment looks wrong; this can definitely happen and will happen most of the time (e.g. if the user never used `define_metric` in the first place)